kayenta-graphite/src/main/java/com/netflix/kayenta/graphite/metrics/GraphiteMetricsService.java

```diff
@@ -0,0 +1,202 @@
+        GraphiteCanaryMetricSetQueryConfig queryConfig =
+                (GraphiteCanaryMetricSetQueryConfig) canaryMetricConfig.getQuery();
+
+        String query = queryConfig.getMetricName();
+        if (!Strings.isNullOrEmpty(canaryScope.getScope())) {
+            query = query.replace(SCOPE_VARIABLE, canaryScope.getScope());
+        }
+
+        if (!Strings.isNullOrEmpty(canaryScope.getLocation())) {
+            query = query.replace(LOCATION_VARIABLE, canaryScope.getLocation());
+        }
+
+        log.debug("Query sent to graphite: {}.", query);
+
+        List<GraphiteResults> graphiteResultsList = remoteService.rangeQuery(
+                query,
+                canaryScope.getStart().getEpochSecond(),
+                canaryScope.getEnd().getEpochSecond(),
+                DEFAULT_FORMAT
+        );
+
+        List<MetricSet> metricSets = new ArrayList<>();
+
+        for (GraphiteResults entry : graphiteResultsList) {
+            metricSets.add(
+                    MetricSet.builder()
+                            .name(canaryMetricConfig.getName())
+                            .startTimeMillis(entry.getStartMills())
+                            .startTimeIso(Instant.ofEpochSecond(entry.getStart()).toString())
+                            .endTimeMillis(entry.getEndMills())
+                            .endTimeIso(Instant.ofEpochSecond(entry.getEnd()).toString())
+                            .stepMillis(entry.getIntervalMills())
+                            .values(entry.getDataPoints().collect(Collectors.toList()))
+                            .build()
+            );
+        }
+
+        return metricSets;
+    }
+
+    @Override
+    public List<Map> getMetadata(String metricsAccountName, String filter) throws IOException {
+        log.debug(String.format("Getting metadata for %s with filter %s", metricsAccountName, filter));
+
+        String baseFilter = "";
+        if (filter.contains(DELIMITER)) {
+            baseFilter = filter.substring(0, filter.lastIndexOf(DELIMITER) + 1);
+        }
+
+        List<Map> result = new LinkedList<>();
+
+        boolean needSpecialDescriptors = filter.contains(DELIMITER)
+                && filter.substring(filter.lastIndexOf(DELIMITER)).contains("$");
+
+        if (needSpecialDescriptors) {
+            result.addAll(getSpecialMetricDescriptors(baseFilter));
+        } else {
+            GraphiteNamedAccountCredentials accountCredentials =
+                    (GraphiteNamedAccountCredentials) accountCredentialsRepository.getOne(metricsAccountName)
+                            .orElseThrow(() -> new IllegalArgumentException(
+                                    String.format("Unable to resolve account %s.", metricsAccountName)));
+
+            GraphiteRemoteService remoteService = accountCredentials.getGraphiteRemoteService();
+
+            filter = convertFilterToGraphiteQuery(filter);
+
+            GraphiteMetricDescriptorsResponse graphiteMetricDescriptorsResponse =
```

Comment:
yes, it responds kind of sufficiently well to ui. if there are lots of users, then the response may not be fast enough.
the current problem of caching is that there will not a lots of metric entries in graphite server and when it reaches an amount, the server will not respond the get all entries request.  so we have to use query level by level to get the typeahead working.  for example the query is a.b.c.d.e , when user types a. the service will only return a.b. 

cache the last response is a good idea, could we leave a TODO here in case we hit performance issue ?

Reply:
Yes, sounds good to me.